pkg/controller/daemon/daemon_controller.go

```diff
@@ -1254,15 +1254,15 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
 	return dsc.updateDaemonSetStatus(ctx, ds, nodeList, hash, true)
 }
 
-// nodeShouldRunDaemonPod checks a set of preconditions against a (node,daemonset) and returns a
+// NodeShouldRunDaemonPod checks a set of preconditions against a (node,daemonset) and returns a
 // summary. Returned booleans are:
 // * shouldRun:
 //     Returns true when a daemonset should run on the node if a daemonset pod is not already
 //     running on that node.
 // * shouldContinueRunning:
 //     Returns true when a daemonset should continue running on a node if a daemonset pod is already
 //     running on that node.
-func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *apps.DaemonSet) (bool, bool) {
+func NodeShouldRunDaemonPod(node *v1.Node, ds *apps.DaemonSet) (bool, bool) {
```

Comment:
if you intend to use this in cluster-autoscaler, can you consider moving it to https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-helpers/apps?

It sounds look like the `Predicates` function should be exported though. The test framework can be refactored to use NodeShouldRunDaemonPod directly.

Reply:
Right, I know autoscaler already embeds a lot of k/k code.

There is no explicit deprecation policy for the component-helpers code, but we are more cautious on the dependencies they have. And it's location will make people think twice before changing the function prototype.

Also, there are active efforts to move anything that pkg/scheduler/framework depends on to component-helpers.
So this location should benefit you in the long run.

Reply:
The code in `component-helpers` can have dependencies only on `k8s.io/apimachinery`, `k8s.io/api` and `k8s.io/client-go`. The functions here have transitive dependencies on `k8s.io/kubernetes`, so I don't think we are able to do that without a more major refactor.

Regarding the `Predicates` function, I've removed its use from e2e framework and instead of that used `NodeShouldRunDaemonPod`. I'm hesitant though to make this function private, as there may be more external dependencies on it. I'd propose to do this in another PR at least, to avoid having to revert the whole change.

Reply:
can you make `Predicates` unexported?

Reply:
Sure, done.